<template>
  <!-- #ifdef APP -->
  <scroll-view style="flex:1">
    <!-- #endif -->
    <view>
      <page-head :title="title"></page-head>
      <view class="uni-padding-wrap uni-common-mt">
        <view class="uni-btn-v uni-common-mt">
          <button type="primary" @tap="addTodo">添加Todo</button>
        </view>
        <view class="uni-btn-v uni-common-mt">
          <button type="primary" @tap="randomFail">随机触发失败重试</button>
        </view>
        <view class="uni-btn-v uni-common-mt">
          <button type="primary" @tap="fail">云对象失败调用</button>
        </view>
        <view class="uni-btn-v uni-common-mt">
          <button type="primary" @tap="success">云对象成功调用</button>
        </view>
      </view>
    </view>
    <!-- #ifdef APP -->
  </scroll-view>
  <!-- #endif -->
</template>

<script>
  export default {
    data() {
      return {
        title: '请求云对象'
      }
    },
    methods: {
      addTodo() {
        const todo = uniCloud.importObject('todo')
        const title = '学习编程'
        const content = '熟悉uts语法'
        todo.add(title, content).then<void>((res: UTSJSONObject) : void => {
          uni.showModal({
            title: '错误',
            content: res['showMessage'] as string,
            showCancel: false
          })
        }).catch<void>((err : any | null) : void => {
          const error = err as UniCloudError
          console.error(error)
        })
      },
      randomFail() {
        const todoObj = uniCloud.importObject('todo', {
          errorOptions: {
            retry: true
          }
        })
        todoObj.randomFail().then<void>((res: UTSJSONObject) : void => {
          uni.showModal({
            title: '错误',
            content: res['showMessage'] as string,
            showCancel: false
          })
        }).catch<void>((err : any | null) : void => {
          const error = err as UniCloudError
          console.error(error)
        })
      },
      fail() {
        const todo = uniCloud.importObject('todo')
        todo.fail().then<void>((res: UTSJSONObject) : void => {
          uni.showModal({
            title: '错误',
            content: 'todo.fail应调用失败，此处错误的触发了成功回调',
            showCancel: false
          })
          console.log('todo.fail: ', res);
        }).catch<void>((err : any | null) : void => {
          const error = err as UniCloudError
          console.error(error)
        })
      },
      success() {
        const todo = uniCloud.importObject('todo')
        todo.success().then<void>((res: UTSJSONObject) : void => {
          uni.showModal({
            title: '错误',
            content: res['showMessage'] as string,
            showCancel: false
          })
        }).catch<void>((err : any | null) : void => {
          const error = err as UniCloudError
          console.error(error)
        })
      }
    }
  }
</script>

<style>

</style>
